test/requestinterception-experimental.spec.ts

```diff
@@ -0,0 +1,854 @@
+          if (request.url().endsWith('.css')) actionResults.push('abort');
+        });
+
+        const response = await (async () => {
+          if (expectedAction === 'continue') {
+            const [serverRequest, response] = await Promise.all([
+              server.waitForRequest('/one-style.css'),
+              page.goto(server.PREFIX + '/one-style.html'),
+            ]);
+            actionResults.push(serverRequest.headers.xaction as ActionResult);
+            return response;
+          } else {
+            return await page.goto(server.PREFIX + '/one-style.html');
+          }
+        })();
+
+        expect(actionResults.length).toBe(1);
+        expect(actionResults[0]).toBe(expectedAction);
+        expect(response.ok()).toBe(true);
+      });
+    }
+
+    it('should intercept', async () => {
+      const { page, server } = getTestState();
+
+      await page.setRequestInterception(true);
+      page.on('request', (request) => {
+        if (utils.isFavicon(request)) {
+          request.continue({}, 0);
+          return;
+        }
+        expect(request.url()).toContain('empty.html');
+        expect(request.headers()['user-agent']).toBeTruthy();
+        expect(request.method()).toBe('GET');
+        expect(request.postData()).toBe(undefined);
+        expect(request.isNavigationRequest()).toBe(true);
+        expect(request.resourceType()).toBe('document');
+        expect(request.frame() === page.mainFrame()).toBe(true);
+        expect(request.frame().url()).toBe('about:blank');
+        request.continue({}, 0);
+      });
+      const response = await page.goto(server.EMPTY_PAGE);
+      expect(response.ok()).toBe(true);
+      expect(response.remoteAddress().port).toBe(server.PORT);
+    });
+    it('should work when POST is redirected with 302', async () => {
+      const { page, server } = getTestState();
+
+      server.setRedirect('/rredirect', '/empty.html');
+      await page.goto(server.EMPTY_PAGE);
+      await page.setRequestInterception(true);
+      page.on('request', (request) => request.continue({}, 0));
+      await page.setContent(`
+        <form action='/rredirect' method='post'>
+          <input type="hidden" id="foo" name="foo" value="FOOBAR">
+        </form>
+      `);
+      await Promise.all([
+        page.$eval('form', (form: HTMLFormElement) => form.submit()),
+        page.waitForNavigation(),
+      ]);
+    });
+    // @see https://github.com/puppeteer/puppeteer/issues/3973
+    it('should work when header manipulation headers with redirect', async () => {
+      const { page, server } = getTestState();
+
+      server.setRedirect('/rrredirect', '/empty.html');
+      await page.setRequestInterception(true);
+      page.on('request', (request) => {
+        const headers = Object.assign({}, request.headers(), {
+          foo: 'bar',
+        });
+        request.continue({ headers }, 0);
+
+        expect(request.continueRequestOverrides()).toEqual({ headers });
```

Comment:
shouldn't you verify this code is called at all?

Reply:
I think this tests primary intention is to check that the navigation does not time out.
Adding a suggestion with a comment.